compiler/rustc_middle/src/mir/query.rs

```diff
@@ -358,13 +358,57 @@ pub enum ConstraintCategory {
     Internal,
 }
 
-#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Hash)]
+// This is needed because `HirId` does not implement `Ord`.
+impl Ord for ConstraintCategory {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        let s = match self {
+            Self::ClosureUpvar(hir_id) => Some(hir_id.index()),
+            _ => None,
+        };
+        let o = match other {
+            Self::ClosureUpvar(hir_id) => Some(hir_id.index()),
+            _ => None,
+        };
+
+        s.cmp(&o)
+    }
+}
+
+impl PartialOrd for ConstraintCategory {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(&other))
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
 #[derive(TyEncodable, TyDecodable, HashStable)]
 pub enum ReturnConstraint {
     Normal,
     ClosureUpvar(hir::HirId),
 }
 
+// This is needed because `HirId` does not implement `Ord`.
+impl Ord for ReturnConstraint {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        let s = match self {
+            &Self::Normal => None,
+            &Self::ClosureUpvar(h) => Some(h.index()),
+        };
+        let o = match other {
+            &Self::Normal => None,
+            &Self::ClosureUpvar(h) => Some(h.index()),
+        };
+
+        s.cmp(&o)
+    }
+}
```

Comment:
Same here.

Could these `Ord` implementations, if they're incorrect, be causing the changes in the UI tests?